tla/etcdraft.tla

```diff
@@ -248,32 +241,71 @@ InitLeaderVars == /\ matchIndex = [i \in Server |-> [j \in Server |-> 0]]
                   /\ pendingConfChangeIndex = [i \in Server |-> 0]
 InitLogVars == /\ log          = [i \in Server |-> <<>>]
                /\ commitIndex  = [i \in Server |-> 0]
+               /\ applied     = [i \in Server |-> 0]
 InitConfigVars == /\ config = [i \in Server |-> [ jointConfig |-> <<InitServer, {}>>, learners |-> {}]]
-                  /\ reconfigCount = 0 
-InitDurableState == 
-    durableState = [ i \in Server |-> [
-        currentTerm |-> currentTerm[i],
-        votedFor |-> votedFor[i],
-        log |-> Len(log[i]),
-        commitIndex |-> commitIndex[i],
-        config |-> config[i]
-    ]]
+                  /\ appliedConfChange = [i \in Server |-> 0] 
+
+InReadyPhase(i) == ready[i] /= <<>>
+
+InitReadyVars == 
+    /\ durableState = [ i \in Server |-> [
+            currentTerm |-> currentTerm[i],
+            votedFor |-> votedFor[i],
+            log |-> Len(log[i]),
+            applied |-> 0,
+            commitIndex |-> commitIndex[i],
+            config |-> config[i]
+        ]]
+    /\ ready = [i \in Server |-> <<>>]
 
 Init == /\ InitMessageVars
         /\ InitServerVars
         /\ InitCandidateVars
         /\ InitLeaderVars
         /\ InitLogVars
         /\ InitConfigVars
-        /\ InitDurableState
+        /\ InitReadyVars
 
 ----
 \* Define state transitions
 
+SpecActions == {
+    "RequestVote",
+    "BecomeLeader",
+    "ClientRequest",
+    "AdvanceCommitIndex",
+    "AppendEntries",
+    "AppendEntriesToSelf",
+    "Heartbeat",
+    "SendSnapshot",
+    "Receive",
+    "Timeout",
+    "StepDownToFollower",
+    "Ready",
+    "PersistState",
+    "ApplyConfChange",
+    "SendMessages",
+    "Advance",
+    "Restart",
+    "DuplicateMessage",
+    "DropMessage",
+    "AddNewServer",
+    "AddLearner",
+    "DeleteServer"
+}
+
+IsEnabled(action, args) == action \in SpecActions
```

Comment:
`SpecActions`, `IsEnabled`, `PostAction` are orthogonal to how the raft algorithm works.  Their purpose is to constrain model checking with TLC.  Including them in the `etcdraft` specification makes the spec less readable.  For example, it would complicate Apalache-based model checking or proving the specification with TLAPS, should either be attempted in the future. Therefore, I strongly recommend limiting the constraints for model checking to the Trace* and MC* specifications.

Reply:
Trace* and MC* specs utilize the same mechanism of constraint in model checking. That's why I put these operators in their common base spec. In such a way we only need to focus on implementing custom IsEnabled and PostAction for different purposes. Or should I make a middle layer spec Constraint*?

Reply:
Decouple the SpecAction definitions into a new spec etcdraft_control.tla. MC* and Trace* specs extends etcdraft_control to implement their own IsEnabled and PostAction operators.